src/core/algorithms/fd/eulerfd/eulerfd.cpp

```diff
@@ -0,0 +1,427 @@
+    auto &tree = negative_cover_[rhs];
+    for (auto &invalid : neg_cover_vector) {
+        if (!invalid[rhs]) {
+            AddInvalidAtTree(tree, invalid);
+        }
+    }
+
+    std::vector<Bitset> neg_cover_rhs;
+    tree.ForEach([&neg_cover_rhs](Bitset const &invalid) { neg_cover_rhs.push_back(invalid); });
+    return neg_cover_rhs;
+}
+
+size_t EulerFD::Invert(size_t rhs, std::vector<Bitset> const &neg) {
+    auto &tree = positive_cover_[rhs];
+    for (auto &invalid : neg) {
+        auto removeds = RemoveGeneralizations(tree, invalid);
+        for (auto &removed : removeds) {
+            for (size_t i = 0; i < number_of_attributes_; i++) {
+                if (i == rhs || invalid[i] || constant_columns_[attribute_indexes_[i]]) {
+                    continue;
+                }
+                Bitset copy = removed;
+                copy.set(i);
+                if (!tree.ContainsAnySubsetOf(copy)) {
+                    tree.Add(copy);
+                }
+            }
+        }
+    }
+    return tree.GetCardinality();
+}
+
+size_t EulerFD::GenerateResults() {
+    // check is new non fd discovered
+    if (old_invalid_size_ == invalids_.size()) {
+        return fd_num_;
+    }
+    old_invalid_size_ = invalids_.size();
+
+    std::vector<Bitset> neg_cover_vector(new_invalids_.begin(), new_invalids_.end());
+    auto new_indexes = GetAttributesSortedByFrequency(neg_cover_vector);
+
+    // check is rebuild of covers neccesary
+    if (attribute_indexes_ != new_indexes) {
+        if (old_invalid_size_ != 0) {
+            InitCovers();
+            neg_cover_vector.assign(invalids_.begin(), invalids_.end());
+            new_indexes = GetAttributesSortedByFrequency(neg_cover_vector);
+        }
+        attribute_indexes_ = new_indexes;
+    }
+
+    std::vector<size_t> inv_indexes(number_of_attributes_);
+    for (size_t i = 0; i < number_of_attributes_; ++i) {
+        inv_indexes[attribute_indexes_[i]] = i;
+    }
+
+    // change attribute order, in first index the largset number of 1
+    for (auto &neg_cover_el : neg_cover_vector) {
+        neg_cover_el = ChangeAttributesOrder(neg_cover_el, inv_indexes);
+    }
+
+    // sorting all non fd by number of 1
+    std::sort(neg_cover_vector.begin(), neg_cover_vector.end(),
+              [](Bitset const &left, Bitset const &right) { return left.count() > right.count(); });
+
+    // creating ncover and pcover trees for each rhs
+    size_t fd_num = 0;
+    for (size_t rhs = 0; rhs < number_of_attributes_; rhs++) {
+        if (constant_columns_[rhs]) {
+            continue;
+        }
+
+        size_t real_rhs = inv_indexes[rhs];
+        auto neg = CreateNegativeCover(real_rhs, neg_cover_vector);
+        // sorting all non fd by number of 1
+        std::sort(neg.begin(), neg.end(), [](Bitset const &left, Bitset const &right) {
+            return left.count() > right.count();
+        });
+        fd_num += Invert(real_rhs, neg);
```

Comment:
```suggestion
        fd_num += Invert(real_rhs, std::move(neg));
```

Reply:
Although the compiler will probably figure out this optimization itself, it is better to write it explicitly.

Reply:
Invert takes a ```const vector &```,  doesn't copy it, only once iterates over it. I don't see any reason for call a move constructor instead pass by reference.